### ISSUE TITLE
Poor performance of varied query

### ISSUE THREAD

**Issue by adamretter** [eXist-db/exist#3418]
The following query executes in:

1. ~5 seconds in BaseX (9.2.4)
2. ~5.5 seconds in Saxon (EE 9.8.0.12)
3. ~18 seconds in eXist-db (5.3.0-SNAPSHOT)

```xquery
let $json := fn:unparsed-text("https://api.github.com/repos/exist-db/exist/stats/contributors")
let $data := fn:json-to-xml($json)
return
<results>{
  for $year in (2005 to 2020)
  let $ut-start := (xs:dateTime($year || "-01-01T00:00:00-00:00") - xs:dateTime("1970-01-01T00:00:00-00:00")) div xs:dayTimeDuration('PT1S')
  let $ut-end := (xs:dateTime(($year + 1)|| "-01-01T00:00:00-00:00") - xs:dateTime("1970-01-01T00:00:00-00:00")) div xs:dayTimeDuration('PT1S')
  order by $year descending
  return
    <year number="{$year}">
      {
        for $user-data in $data/fn:array/fn:map
        let $username := $user-data//fn:map[@key eq "author"]/fn:string[@key eq "login"]/string(.)
        let $commits := fn:sum($user-data/fn:array[@key eq "weeks"]/fn:map[fn:number[@key eq "w"][xs:int(.) ge $ut-start][xs:int(.) lt $ut-end]]/fn:number[@key eq "c"]/xs:int(.))
        where $commits gt 0
        order by $commits descending
        return
        <user name="{$username}">{$commits}</user>
          
      }
    </year>
}</results>
```

**Comment by line-o** [comment]
Where is the bug? @adamretter

**Comment by adamretter** [comment]
> Would you consider the runtime of this query a bug, @adamretter ?

No. I am not reporting a bug. I am reporting a performance issue. I believe the profile and problem is different from the many problems illustrated in #3406... otherwise I would not of posted it ;-)

The problem is not the downloading of an external resource. Previously I was reading it directly from the local filesystem, I rather added that to make it portable between implementations and people. I should have perhaps mentioned it.

Sure, the problem needs to be broken down and investigated, but the first step is always reporting that there is a problem. I didn't want the issue to be lost.

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

Returning to this 2020 issue. I synthesized a payload mimicking the GitHub `/stats/contributors` API (N users × W weeks per user) and re-ran the reproducer's structure on develop tip `b917e1ab1d` (eXist 6.x), JDK Zulu 21.38.21:

| users | weeks | years | innerOnly ms | as written ms | hoist `$data/fn:array/fn:map` ms | per-user fully hoisted ms | speedup |
|------:|------:|------:|-------------:|--------------:|---------------------------------:|--------------------------:|--------:|
| 10    | 52    | 16    | 18           | 71            | 55                               | 30                        | 2.37×   |
| 30    | 52    | 16    | 13           | 85            | 82                               | 68                        | 1.25×   |
| 50    | 52    | 16    | 15           | 127           | 127                              | 107                       | 1.19×   |
| 50    | 200   | 16    | 53           | 502           | 496                              | 412                       | 1.22×   |

Two observations:

1. **The original 18s timing is no longer reproducible.** At a workload approximating the GitHub stats API shape (50 users × 52 weeks × 16 years), eXist now runs the reproducer in ~127ms — two orders of magnitude faster than the 2020 number. Six years of memtree, json-to-xml, datetime-arithmetic, and primitive-cast improvements have closed the bulk of the original gap.

2. **Manually hoisting per-user invariants outside the year loop saves only ~20% at issue-relevant scale.** Lifting `$data/fn:array/fn:map` to a `let` saves nothing measurable (memtree path traversal is fast). Lifting both `$username` and the weeks-array reference saves 18-22%. The remaining cost is intrinsic: 800 invocations of `fn:sum(...)` filtering 52 weeks each, where each year's filter is different and cannot be hoisted.

The 3.5× gap to BaseX from 2020 likely came from BaseX's faster primitives (datetime arithmetic, integer casts, in-memory path traversal) rather than an algorithmic optimizer-level difference. Closing the remaining ~20% on eXist would require loop interchange + per-user memoization (i.e., a full FLWOR fixpoint pipeline modeled on BaseX's `GFLWOR.optimize()`), which is being tracked as a separate audit recommendation.

A side finding: a recent local experimental branch attempted exactly the loop-invariant input hoist this issue suggested (rewrite `for $u in $data/fn:array/fn:map` to `let $users := $data/fn:array/fn:map; for $u in $users`). It **regressed** the reproducer by 7-21% — the analysis machinery adds compile-time cost, but the hoisted path traversal it would skip is already cheap. So the obvious rewrite is empirically counter-productive on this shape.

Suggesting we leave this open pending an apples-to-apples re-run against current BaseX (12.x) on the original `unparsed-text("https://api.github.com/...")` reproducer. If the gap is now within ~30%, this can probably be closed.

Diagnostic harness (`Issue3418Benchmark.java`) and a longer-form design write-up are available on request.

**Comment by duncdrum** [comment]
@joewiz if you feel there is need for an up-to-date comparison with BaseX or Saxon, go ahead, otherwise I propose to close this.

**Comment by joewiz** [comment]
[This response was prompted by Joe, drafted by Claude Code, and reviewed by Joe.]

@duncdrum I ran the comparison — here it is, so the close call can be made against current numbers.

The reproducer is pure XQuery 3.1 (`fn:json-to-xml`, nested FLWOR, `fn:sum`) with no eXist extensions, so it runs unmodified everywhere. I checked that BaseX and Saxon produce identical output before timing anything.

All three engines, same machine, today. eXist is current `develop` (`61a9b9ebac`), against BaseX 12.4 and Saxon-HE 12.5. Payload is the GitHub `/stats/contributors` shape from the original report: N users, each with a `weeks` array, aggregated per year across 2005–2020.

| Scale | eXist | BaseX 12.4 | BaseX eval only | Saxon 12.5 eval |
|---|---|---|---|---|
| 10 users × 52 weeks × 16y | 99–124 ms | 59 ms | 18 ms | 13 ms |
| 30 × 52 × 16y | 107–130 ms | 79 ms | 38 ms | 37 ms |
| **50 × 52 × 16y** (issue scale) | **172–223 ms** | **91 ms** | 47 ms | 40 ms |
| 50 × 200 × 16y | 733–782 ms | 208 ms | 153 ms | 134 ms |

Reading the columns: the eXist and BaseX columns both cover parse + compile + evaluate. Saxon's `-repeat` compiles once and re-executes, so its column is evaluation-only and is not comparable to the first two — BaseX's eval-only column is there to sit beside it. The eXist figures are ranges across two full runs; run-to-run variance is 20–30%, so point values would overstate the precision.

Two conclusions, and they point in different directions:

**The reported pathology is gone.** The 2020 report was 18 seconds on 5.3.0-SNAPSHOT. Issue scale now runs in about 0.2 s — roughly 80× faster. Whatever caused that was fixed years ago.

**A gap remains.** eXist is about 2× BaseX at issue scale, widening to roughly 3.5–3.8× at four times the data. So "the 18 s is unreproducible" and "there is nothing left here" are separate claims, and only the first holds.

On whether it is fixable cheaply: I don't think it is. When I diagnosed this in May I tested manual hoisting of the loop-invariant parts — binding `$data/fn:array/fn:map` outside the year loop, and a fuller per-user hoist of username and weeks. Across the runs above those variants land between 0.71× and 1.21×, sometimes *slower* than the query as written. The residual cost is re-filtering 52 weeks per user per year, and each year's filter differs, so there is no loop-invariant work for an optimizer to lift. Closing it would need loop interchange plus per-user memoization — a real FLWOR fixpoint pipeline, not a targeted rewrite.

So I'm happy for this to be closed, but I'd suggest closing it on the accurate grounds: the reported regression is long fixed, and what remains is a ~2× constant factor on one query shape that is not worth a fixpoint optimizer on its own. That is a different statement from eXist having caught up, and it seems worth having on the record in case someone benchmarks this shape again.

If anyone wants to re-run or extend this, the harness that produced the table is at [`bench/3418-engine-comparison` on my fork](https://github.com/joewiz/exist/tree/bench/3418-engine-comparison) — current `develop` plus one test file, [`Issue3418Benchmark.java`](https://github.com/joewiz/exist/blob/fef2d9dc3d741f227c9af441be206c506cae833b/exist-core/src/test/java/org/exist/xquery/Issue3418Benchmark.java). Run it with:

```bash
mvn test -pl exist-core -Dtest=Issue3418Benchmark -Dexist.run.benchmarks=true
```

It synthesizes the same payload the numbers above came from, and its `reproducerQuery` emits plain XQuery 3.1 you can drop straight into BaseX or Saxon for the cross-engine columns. The branch is a diagnostic artifact, not a merge candidate.
